### PR TITLE
BUG: numpy.array does not respect __iter__ = None

### LINKED ISSUE #27382: BUG: numpy.array does not respect `__iter__ = None`
### Describe the issue:

Since python3 `__iter__ = None` means, this object is not iterable.
Creating an array of `dtype=object` with such objects leads to a `TypeError`

I would expect, that numpy doesn't investigate these objects any deeper, and simply puts them into the created array,

### Reproduce the code example:

```python
import numpy as np

class SomeClass:
    def __init__(self, size):
        self.size = size

    __iter__ = None  # this object is not iterable

    def __getitem__(self, index):
        if index < self.size:
            return index ** 2
        raise IndexError(index)

    def __len__(self):
        return self.size

array = np.array([SomeClass(5), SomeClass(7)], dtype=object)
```

### Error message:

```shell
Traceback (most recent call last):
  File "./iter.py", line 17, in <module>
    array = np.array([SomeClass(5), SomeClass(7)], dtype=object)
TypeError: Could not convert object to sequence
```

### Python and NumPy Versions:

2.0.2
3.9.16 (main, Mar  8 2023, 14:00:05) 
[GCC 11.2.0]

### Runtime Environment:

_No response_

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by ruema**
This PR allow, that classes define `__iter__ = None` to prevent array coercion.
Fixes #27382.

### CODE REVIEW

**Comment by melissawm** [comment]
Hi @ruema, thanks for the PR! 

If your work fixes the issue you mentioned, please use [the GitHub keywords to link this PR to that issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). 

Also check out our [contributor guide](https://numpy.org/devdocs/dev/index.html) and [instructions on how to write the commit message](https://numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message). Feel free to ping me if you have any questions!

**Comment by ruema** [comment]
Hi melissawm,

thank you for your remarks. I've updated the commit.

**Comment by seberg** [comment]
@ruema this definitely needs a test, otherwise LGTM on first sight.

Do you know why `PySequence_Check` does not check for this for us, though?!  Maybe we should open a Python issue for that?  Or is this *backporting* that behavior to older Python versions I.e. `PySequence_Check` does this for us on new Python versions, but you want this to work on older ones as well?

If that is the case, I think it would be nice to make that explicit in this PR, either by moving the check to the `PySequence_Check` (with a version dependency), or just adding an `#ifdef` with a comment here.

**Comment by ruema** [comment]
@seberg: I didn't find tests for dict like objects and sequences either. So I added one.
The documentation to PySequence_Check states, that any object with a __getitem__ method may be a sequence, but this is not a definite test.
So I think, each library, that uses this function has to handle the corner cases by itself.
But it is definitely worth discussing, to add more checks to PySequence_Check.

**Comment by seberg** [comment]
>  I didn't find tests for dict like objects and sequences either.

Thanks for the update!  There probably some in another part of the tests, the array-coercion test file was added much later.  But having more here is good (I'll look closer at it later).

> that any object with a `__getitem__` method may be a sequence, but this is not a definite test.

Right, I still suspect we should add this check together with the `PySequence_Check`, effectively, this changes the definition of what we consider a "sequence".
At least I think this basically is "sequence that is also iterable" and the iterable check should be fast without the effectively  try/except.

I think I'll tag this for triage, I am now slightly worried that definition change has some subtle other behavior impliciations, although I think it just turns an error into a success.

There is also the subtlety it is a source of confusion to users that some objects get added as scalar elements.  There are two cases that are particularly odd/confusing:
* Iterables that are not sequences.  (and to a lesser degree dicts also)
* SciPy sparse arrays/matrices look a lot like arrays, but are also not sequences (for some reason).  Oddly/worse, that makes some things work, that shouldn't otherwise work...

So it would be nice to consider briefly that this wouldn't broaden such confusion.

**Comment by seberg** [comment]
I brought it up yesterday a bit.  I don't think anyone could think of a problme with doing this, so unless one jumps up, I think we can do this eventually.
I still have two questions though:
1. I do think the check for `__iter__ = None` seems off here.  It should be a check for "iterable", I believe (presumably that means the slot is defined, I _suspect_ `__iter__ = None` just prevents auto-filling it at class creation).
   And since I suspect that check is very light-weight, I also think it would probably pair better with the `PySequence_Check` (and be much clearer there).
2. To help grease the decision a bit:  Do you have a use-case/example for an object where you need this?

**Comment by 97gamjak** [comment]
Hi @seeberg,

@galjos and I stumbled across this PR and the related issue today and might have and actual use case for this.

so basically what we have is a container class that is completely unordered, however, we need `__getitem__` for some kind of selection. At first our class was somehow defined like this, which worked perfectly fine with `numpy.atleast_1d`:

```
class Foo():
    def __getitem__(self,key):
        return 0
```

then we wanted to add the `__len__` function to this class and then `numpy.atleast_1d(Foo())` ends in an endless loop. Afterwards, we then tried to delete the `__iter__` function like this:

```
class Foo():
    def __getitem__(self,key):
        return 0
    def __len__(self):
        return 0
    __iter__ = None
```

which results in the following error in the current release of `2.2.1` - `Could not convert object to sequence`, when calling `numpy.atleast_1d(Foo())`

Now, this pull request fixes the issue but however as soon as we try to call `numpy.atleast_1d([Foo(),Foo()])` on a list for example which contains at least two instances of the object `Foo` we get following error:

```
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.
```

TLDR: in our use case this PR solves the primary exception that was raised when deleting the `__iter__` method, but however, now it results in some unexpected behaviour with other arguments to `numpy.atleast_1d` (at least in our opinion)

I hope i was able to describe our issue clearly enough :D

**Comment by seberg** [comment]
@97gamjak yeah, just hearing that you have something where you would use this at all helps, thanks.

I suspect the main question is if anyone has a problem with implicitly creating the `object` dtype array here.  IIRC, I thought this seems not so bad, because the objects are not iterable, and it is much like adding dicts (and not so much like sparse arrays).

> on a list for example which contains at least two instances of the object Foo we get following error:

That seems like a bug in the current code.
